### ISSUE TITLE
How async mode works?

### ISSUE THREAD

**Issue by zdm** [Azq2/perl-html5-dom#2]
Hi,
It is unclear, how async mode works. For example in the following code I am expecting, that printed value will be "0". Because $html is pretty large and parsed in async mode. But it is always "1". Why?
```
my $parser = HTML5::DOM->new( {
    threads => 4,
    async   => 1,
} );
my $tree = $parser->parse($html);
say $tree->parsed ? 1 : 0;
```

**Comment by Azq2** [comment]
Hello!
This is a bug. I misunderstood how myhtml_parse works.
Now i'm trying to fix it and proide real async mode.

**Comment by Azq2** [comment]
I added parseAsync function. Check docs for 1.09 (see examples). 
Also it supports built-in integration with EV or any custom event-loop using pipes/socketpair. 
1.09 available on metacpan after few hours (reindexing is slow).

Please, test it and close this issue if problem resolved.

Thanks for feedback!

**Comment by zdm** [comment]
Perfect, I will test at the evening and will write you.

**Comment by Azq2** [comment]
I have same perfomance degradation on original Modest C source.

```
$ bin/myhtml/print_tree_high_level ~/index.html                                                                                                                                     
1 threads: 635.027258
2 threads: 674.421497
3 threads: 663.557751
```

My friends tested on AMD Ryzen and intel i7/xeon. Same results - single thread faster. 

"One click" test:
```
git clone https://github.com/lexborisov/Modest
cd Modest
curl https://html.spec.whatwg.org/ > ~/index.html
curl https://dumpz.org/aNdawx3GKn3Q/text/ > ./examples/myhtml/print_tree_high_level.c
make
bin/myhtml/print_tree_high_level ~/index.html
```

Test code:
```C
#include <stdio.h>
#include <stdlib.h>
#include <myhtml/api.h>
#include <sys/time.h>

struct res_html {
	char	*html;
	size_t	size;
};

struct res_html load_html_file(const char* filename)
{
	FILE *fh = fopen(filename, "rb");
	if(fh == NULL) {
		fprintf(stderr, "Can't open html file: %s\n", filename);
		exit(EXIT_FAILURE);
	}
	
	if(fseek(fh, 0L, SEEK_END) != 0) {
		fprintf(stderr, "Can't set position (fseek) in file: %s\n", filename);
		exit(EXIT_FAILURE);
	}
	
	long size = ftell(fh);
	
	if(fseek(fh, 0L, SEEK_SET) != 0) {
		fprintf(stderr, "Can't set position (fseek) in file: %s\n", filename);
		exit(EXIT_FAILURE);
	}
	
	if(size <= 0) {
		fprintf(stderr, "Can't get file size or file is empty: %s\n", filename);
		exit(EXIT_FAILURE);
	}
	
	char *html = (char*)malloc(size + 1);
	if(html == NULL) {
		fprintf(stderr, "Can't allocate mem for html file: %s\n", filename);
		exit(EXIT_FAILURE);
	}
	
	size_t nread = fread(html, 1, size, fh);
	if (nread != size) {
		fprintf(stderr, "could not read %ld bytes (%ld bytes done)\n", size, nread);
		exit(EXIT_FAILURE);
	}
	
	fclose(fh);
	
	struct res_html res = {html, (size_t)size};
	return res;
}

double current_timestamp()
{
	struct timeval t;
	gettimeofday(&t, NULL);
	return (double) t.tv_sec * 1000.0 + (double) t.tv_usec / 1000.0;
}

int main(int argc, const char * argv[])
{
	const char* path;
	
	if (argc == 2) {
		path = argv[1];
	}
	else {
		printf("Bad ARGV!\nUse: print_tree_high_level <path_to_html_file>\n");
		exit(EXIT_FAILURE);
	}
	
	int tries = 20;
	double start, elapsed;
	
	struct res_html res = load_html_file(path);
	
	myhtml_t* myhtml = myhtml_create();
	myhtml_init(myhtml, MyHTML_OPTIONS_PARSE_MODE_SINGLE, 1, 0);
	
	start = current_timestamp();
	
	for (int i = 0; i < tries; ++i) {
		myhtml_tree_t* tree = myhtml_tree_create();
		myhtml_tree_init(tree, myhtml);
		
		myhtml_parse(tree, MyENCODING_UTF_8, res.html, res.size);
		
		myhtml_tree_destroy(tree);
	}
	
	elapsed = current_timestamp() - start;
	
	myhtml_destroy(myhtml);
	
	printf("1 threads: %f\n", elapsed / tries);
	
	myhtml = myhtml_create();
	myhtml_init(myhtml, MyHTML_OPTIONS_DEFAULT, 2, 0);
	
	start = current_timestamp();
	
	for (int i = 0; i < tries; ++i) {
		myhtml_tree_t* tree = myhtml_tree_create();
		myhtml_tree_init(tree, myhtml);
		
		myhtml_parse(tree, MyENCODING_UTF_8, res.html, res.size);
		
		myhtml_tree_destroy(tree);
	}
	
	elapsed = current_timestamp() - start;
	
	myhtml_destroy(myhtml);
	
	printf("2 threads: %f\n", elapsed / tries);
	myhtml = myhtml_create();
	myhtml_init(myhtml, MyHTML_OPTIONS_DEFAULT, 3, 0);
	
	start = current_timestamp();
	
	for (int i = 0; i < tries; ++i) {
		myhtml_tree_t* tree = myhtml_tree_create();
		myhtml_tree_init(tree, myhtml);
		
		myhtml_parse(tree, MyENCODING_UTF_8, res.html, res.size);
		
		myhtml_tree_destroy(tree);
	}
	
	elapsed = current_timestamp() - start;
	
	myhtml_destroy(myhtml);
	
	printf("3 threads: %f\n", elapsed / tries);
	
	return 0;
}
```

Currently i no have idea why. May be:
1. I do something wrong when using myhtml api.
2. Bug on Linux/Windows in myhtml.

I continue trying to understand real cause of degradation.

**Comment by zdm** [comment]
I tested parseAsync with Coro, it works perfectly, thank you.

One more issue - version 1.10 not installed under windows.
```
...
blib\arch')" t/*.t t/*/*.t
t/0-api.t .. Dubious, test returned 5 (wstat 1280, 0x500)
All 213 subtests passed

Test Summary Report
-------------------
t/0-api.t (Wstat: 1280 Tests: 213 Failed: 0)
  Non-zero exit status: 5
  Parse errors: No plan found in TAP output
Files=1, Tests=213,  2 wallclock secs ( 0.03 usr +  0.00 sys =  0.03 CPU)
Result: FAIL
Failed 1/1 test programs. 0/213 subtests failed.
gmake: *** [Makefile:1066: test_dynamic] Error 5
FAIL
...
```

**Comment by Azq2** [comment]
@zdm fixed in 1.11 (tested on Strawberry)

**Comment by Azq2** [comment]
I think problem resolved. Please, open if not.
